docs/JAX_basic_text_classification.md

```diff
@@ -0,0 +1,455 @@
+!echo "Number of positive samples in test set:"
+!ls /tmp/data/imdb/aclImdb/test/pos | wc -l
+!echo "Number of negative samples in test set:"
+!ls /tmp/data/imdb/aclImdb/test/neg | wc -l
+!echo "First 10 files with positive samples in train/test sets:"
+!ls /tmp/data/imdb/aclImdb/train/pos | head
+!ls /tmp/data/imdb/aclImdb/test/pos | head
+!echo "Display a single positive sample:"
+!cat /tmp/data/imdb/aclImdb/train/pos/6248_7.txt
+```
+
+Next, we create the dataset Python class to read samples from the disk, use [Tiktoken](https://github.com/openai/tiktoken) to encode raw text into tokens and use [Grain](https://github.com/google/grain) for efficient data loading and batching.
+
+```{code-cell} ipython3
+from pathlib import Path
+
+
+class SentimentAnalysisDataset:
+    def __init__(self, path: str | Path):
+        self.path = Path(path)
+        assert self.path.exists()
+
+        pos_texts = list((self.path / "pos").glob("*.txt"))
+        neg_texts = list((self.path / "neg").glob("*.txt"))
+        self.text_files = pos_texts + neg_texts
+        assert len(self.text_files) > 0
+        # Label 0 for Positive comments
+        # Label 1 for Negative comments
+        self.labels = [0] * len(pos_texts) + [1] * len(neg_texts)
+
+    def __len__(self) -> int:
+        return len(self.text_files)
+
+    def read_text_file(self, path: str | Path) -> str:
+        with open(path, "r") as handler:
+            lines = handler.readlines()
+        return "\n".join(lines)
+
+    def __getitem__(self, index: int) -> tuple[str, int]:
+        label = self.labels[index]
+        text = self.read_text_file(self.text_files[index])
+        return {"text": text, "label": label}
+
+
+root_path = Path("/tmp/data/imdb/aclImdb/")
+train_dataset = SentimentAnalysisDataset(root_path / "train")
+test_dataset = SentimentAnalysisDataset(root_path / "test")
+
+print("- Number of samples in train and test sets:", len(train_dataset), len(test_dataset))
+print("- First train sample:", train_dataset[0])
+print("- First test sample:", test_dataset[0])
+```
+
+Now, we can create string to tokens preprocessing transformation and set up data loaders. We are going to use the GPT-2 tokenizer via [Tiktoken](https://github.com/openai/tiktoken).
+
+```{code-cell} ipython3
+import numpy as np
+
+import tiktoken
+import grain.python as grain
+
+
+tokenizer = tiktoken.get_encoding("gpt2")
+
+
+class TextPreprocessing(grain.MapTransform):
+    def __init__(self, tokenizer, max_length: int = 256):
+        self.tokenizer = tokenizer
+        self.max_length = max_length
+
+    def map(self, data):
+        text = data["text"]
+        encoded = self.tokenizer.encode(text)
+        # Cut to max length
+        encoded = encoded[:self.max_length]
+        # Pad with zeros if needed
+        encoded = np.array(encoded + [0] * (self.max_length - len(encoded)))
+        return {
+            "text": encoded,
+            "label": data["label"],
+        }
+```
+
```

Comment:
If you are using two separate code cells for these actions, I would keep the grain import in the next cell so they are "self-contained" (they are not really self-contained, but it makes for a better cognitive unit)